datacube/index/memory/_datasets.py

```diff
@@ -532,77 +535,64 @@ def _search(
             if return_format == self.RET_FORMAT_PRODUCT_GROUPED and product_results:
                 yield (product_results, product)
 
-    def _search_flat(
-            self,
-            limit: Optional[int] = None,
-            source_filter: Optional[Mapping[str, QueryField]] = None,
-            archived: bool | None = False,
-            **query: QueryField
-    ) -> Iterable[Dataset]:
-        return cast(Iterable[Dataset], self._search(
-                    return_format=self.RET_FORMAT_DATASETS,
-                    limit=limit,
-                    source_filter=source_filter,
-                    archived=archived,
-                    **query)
-        )
-
-    def _search_grouped(
-            self,
-            limit: Optional[int] = None,
-            source_filter: Optional[Mapping[str, QueryField]] = None,
-            archived: bool | None = False,
-            **query: QueryField
-    ) -> Iterable[Tuple[Iterable[Dataset], Product]]:
-        return cast(Iterable[Tuple[Iterable[Dataset], Product]], self._search(
-                    return_format=self.RET_FORMAT_PRODUCT_GROUPED,
-                    limit=limit,
-                    source_filter=source_filter,
-                    archived=archived,
-                    **query)
-        )
-
     def _get_prod_queries(self, **query: QueryField) -> Iterable[Tuple[Mapping[str, QueryField], Product]]:
         return ((q, product) for product, q in self._index.products.search_robust(**query))
 
-    @deprecat(
-        deprecated_args={
-            "source_filter": {
-                "reason": "Filtering by source metadata is deprecated and will be removed in future.",
-                "version": "1.9.0",
-                "category": ODC2DeprecationWarning
-
-            }
-        }
-    )
-    def search(self,
-               limit: Optional[int] = None,
-               source_filter: Optional[Mapping[str, QueryField]] = None,
-               archived: bool | None = False,
-               **query: QueryField) -> Iterable[Dataset]:
+    def _search(self,
+                limit: Optional[int] = None,
+                source_filter: Optional[Mapping[str, QueryField]] = None,
+                archived: bool | None = False,
+                **query: QueryField) -> Iterable[Dataset]:
         return cast(
             Iterable[Dataset],
-            self._search_flat(limit=limit, source_filter=source_filter, archived=archived, **query)
+            self._search_impl(
+                return_format=self.RET_FORMAT_DATASETS,
+                limit=limit, source_filter=source_filter, archived=archived,
+                **query)
         )
 
-    def search_by_product(self,
+    def _search_by_product(self,
+                           archived: bool | None = False,
+                           fetch_all_per_product: bool = False,
```

Comment:
why are we calling it `fetch_all_per_product` instead of `fetch_all`?

Reply:
Because it's subtly different. search_by_product  has two levels of indirection - it returns an iterator of iterators.  The `fetch_all` argument in the wrapper function handles the outer iterator, this argument handles the inner iterator.